File: 300_Longest_Increasing_Subsequence.java
    /*  Explanation
    # Leetcode problem link : https://leetcode.com/problems/longest-increasing-subsequence/
    Time Complexity for operators : o(n^2) .. n is the length of the ENVELOPS
    Extra Space Complexity for operators : o(n)
    Did this code successfully run on Leetcode : NA
    Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
        # Basic approach : 
        # Optimized approach: 
                              
            # 1. 
                    A) 

    */  

class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums == null || nums.length == 0)
            return 0;
        
        int max = 1;
        int dp[] = new int[nums.length];
        Arrays.fill(dp,1);
        for(int i = 1;i < nums.length; i++){
            for(int j = 0;j<i;j++){
                if(nums[i]>nums[j]){
                    dp[i] = Math.max(dp[i],dp[j]+1);
                    max = Math.max(max,dp[i]);
                }
            }
        }
        return max;
    }
}

File: 300-LongestIncreasingSubsequence.cpp
/* 
    Time Complexity                              :  O(N*N) - dynamicPrograaming
                                                    O(N Log N) - optimized
    Space Complexity                             :  O(N) to store in the dp array
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // return dynamicProgramming(nums);
        return optimized(nums);
    }
    
    int dynamicProgramming(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n,1);
        int mx = 1;
        
        for(int i=1;i<n;i++) {
            for(int j=0;j<i;j++) {
                if(nums[i] > nums[j]) {
                    dp[i] = max(dp[i],dp[j]+1);
                }
            }
            mx = max(mx, dp[i]);
        }
        
        return mx;
    }
    
    int optimized(vector<int>& nums) {
        int n = nums.size(), len=0;
        vector<int> dp(n);
        dp[len] = nums[0];
        for(int i=1;i<n;i++) {
            int csz = dp.size()-1;
            if(nums[i] > dp[len]) {
                dp[len+1] = nums[i];
                len++;
            } else {
                int bsIndex = binarySearch(dp, 0, len, nums[i]);
                // int bsIndex = lower_bound(begin(dp), end(dp), nums[i]) - begin(dp);
                dp[bsIndex] = nums[i];
            }
        }
        
        return len+1;
    }
    
    int binarySearch(vector<int>& nums, int low, int high, int target) {
        
        while(low <= high) {
            int mid = low + (high - low)/2;
            if(nums[mid] == target) {
                return mid;
            } else if(nums[mid] > target) {
                high = mid-1;
            } else {
                low = mid+1;
            }
        }
        
        return low;
    }
};